In [329]:
# Importing required libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
from imblearn.over_sampling import SMOTE

# Data Loading and initial understanding

In [330]:
# Loading dataset
df=pd.read_csv('train.csv')
df_test=pd.read_csv('test.csv')

In [331]:
# Printing 1st 5 rows of train dataset
df.head()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0   0        109             0.0             0.0             0.0   
1   1        109             0.0             0.0             0.0   
2   2        109             0.0             0.0             0.0   
3   3        109             0.0             0.0             0.0   
4   4        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014   31.277   
1            6/30/2014            7/31/2014            8/31/2014    0.000   
2            6/30/2014            7/31/2014            8/31/2014   60.806   
3            6/30/2014            7/31/2014            8/31/2014  156.362   
4            6/30/2014            7/31/2014            8/31/2014  240.708   

    arpu_7  ...  sachet_3g_7  sachet_3g_8  fb_user_6  fb_user_7  fb_user_8  \
0   87.009  ...            0            0        NaN        NaN        NaN   
1  122.787  ...            0            0        NaN        1.0        NaN   
2  103.176  ...            0            0        NaN        NaN        NaN   
3  205.260  ...            0            0        NaN        NaN        NaN   
4  128.191  ...            1            0        1.0        1.0        1.0   

    aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  churn_probability  
0  1958         0.0         0.0         0.0                  0  
1   710         0.0         0.0         0.0                  0  
2   882         0.0         0.0         0.0                  0  
3   982         0.0         0.0         0.0                  0  
4   647         0.0         0.0         0.0                  0  

[5 rows x 172 columns]

In [332]:
# Printing 1st 5 rows of test dataset
df_test.head()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0  69999        109             0.0             0.0             0.0   
1  70000        109             0.0             0.0             0.0   
2  70001        109             0.0             0.0             0.0   
3  70002        109             0.0             0.0             0.0   
4  70003        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014   91.882   
1            6/30/2014            7/31/2014            8/31/2014  414.168   
2            6/30/2014            7/31/2014            8/31/2014  329.844   
3            6/30/2014            7/31/2014            8/31/2014   43.550   
4            6/30/2014            7/31/2014            8/31/2014  306.854   

    arpu_7  ...  sachet_3g_6  sachet_3g_7  sachet_3g_8  fb_user_6  fb_user_7  \
0   65.330  ...            0            0            0        NaN        NaN   
1  515.568  ...            0            0            0        NaN        NaN   
2  434.884  ...            0            0            0        NaN        NaN   
3  171.390  ...            0            0            0        NaN        NaN   
4  406.289  ...            0            0            0        NaN        NaN   

   fb_user_8   aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  
0        NaN  1692        0.00        0.00        0.00  
1        NaN  2533        0.00        0.00        0.00  
2        NaN   277      525.61      758.41      241.84  
3        NaN  1244        0.00        0.00        0.00  
4        NaN   462        0.00        0.00        0.00  

[5 rows x 171 columns]

In [333]:
# checking shape of dataset
print(df.shape)
print(df_test.shape)

(69999, 172)
(30000, 171)


In [334]:
# checking information (d-type, null values) of dataset
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 172 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    id                        int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    arpu_6                    float64
 9    arpu_7                    float64
 10   arpu_8                    float64
 11   onnet_mou_6               float64
 12   onnet_mou_7               float64
 13   onnet_mou_8               float64
 14   offnet_mou_6              float64
 15   offnet_mou_7              float64
 16   offnet_mou_8              float64
 17   roam_ic_mou_6             float64
 18   roam_ic_mou_7             float64
 19   roam_ic_mou_8             float64
 20   roam

In [335]:
# Checking numerical information
df.describe(include='all')

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   69999.000000    69999.0         69297.0         69297.0   
unique           NaN        NaN             NaN             NaN   
top              NaN        NaN             NaN             NaN   
freq             NaN        NaN             NaN             NaN   
mean    34999.000000      109.0             0.0             0.0   
std     20207.115084        0.0             0.0             0.0   
min         0.000000      109.0             0.0             0.0   
25%     17499.500000      109.0             0.0             0.0   
50%     34999.000000      109.0             0.0             0.0   
75%     52498.500000      109.0             0.0             0.0   
max     69998.000000      109.0             0.0             0.0   

        loc_ic_t2o_mou last_date_of_month_6 last_date_of_month_7  \
count          69297.0                69999                69600   
unique             NaN                    1                    1   
top                NaN            6/30/2014            7/31/2014   
freq               NaN                69999                69600   
mean               0.0                  NaN                  NaN   
std                0.0                  NaN                  NaN   
min                0.0                  NaN                  NaN   
25%                0.0                  NaN                  NaN   
50%                0.0                  NaN                  NaN   
75%                0.0                  NaN                  NaN   
max                0.0                  NaN                  NaN   

       last_date_of_month_8        arpu_6        arpu_7  ...   sachet_3g_7  \
count                 69266  69999.000000  69999.000000  ...  69999.000000   
unique                    1           NaN           NaN  ...           NaN   
top               8/31/2014           NaN           NaN  ...           NaN   
freq                  69266           NaN           NaN  ...           NaN   
mean                    NaN    283.134365    278.185912  ...      0.081444   
std                     NaN    334.213918    344.366927  ...      0.634547   
min                     NaN  -2258.709000  -1289.715000  ...      0.000000   
25%                     NaN     93.581000     86.714000  ...      0.000000   
50%                     NaN    197.484000    191.588000  ...      0.000000   
75%                     NaN    370.791000    365.369500  ...      0.000000   
max                     NaN  27731.088000  35145.834000  ...     33.000000   

         sachet_3g_8     fb_user_6     fb_user_7     fb_user_8           aon  \
count   69999.000000  17568.000000  17865.000000  18417.000000  69999.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        0.085487      0.916325      0.909544      0.890319   1220.639709   
std         0.680035      0.276907      0.286842      0.312501    952.426321   
min         0.000000      0.000000      0.000000      0.000000    180.000000   
25%         0.000000      1.000000      1.000000      1.000000    468.000000   
50%         0.000000      1.000000      1.000000      1.000000    868.000000   
75%         0.000000      1.000000      1.000000      1.000000   1813.000000   
max        41.000000      1.000000      1.000000      1.000000   4337.000000   

          aug_vbc_3g    jul_vbc_3g   jun_vbc_3g  churn_probability  
count   69999.000000  69999.000000  69999.00000       69999.000000  
unique           NaN           NaN          NaN                NaN  
top              NaN           NaN          NaN                NaN  
freq             NaN           NaN          NaN                NaN  
mean       68.108597     65.935830     60.07674           0.101887  
std       269.328659    267.899034    257.22681           0.302502  
min         0.0

## Data cleaning

In [336]:
# Checking any presence of duplicate value in datatset
df[df.duplicated()]

Empty DataFrame
Columns: [id, circle_id, loc_og_t2o_mou, std_og_t2o_mou, loc_ic_t2o_mou, last_date_of_month_6, last_date_of_month_7, last_date_of_month_8, arpu_6, arpu_7, arpu_8, onnet_mou_6, onnet_mou_7, onnet_mou_8, offnet_mou_6, offnet_mou_7, offnet_mou_8, roam_ic_mou_6, roam_ic_mou_7, roam_ic_mou_8, roam_og_mou_6, roam_og_mou_7, roam_og_mou_8, loc_og_t2t_mou_6, loc_og_t2t_mou_7, loc_og_t2t_mou_8, loc_og_t2m_mou_6, loc_og_t2m_mou_7, loc_og_t2m_mou_8, loc_og_t2f_mou_6, loc_og_t2f_mou_7, loc_og_t2f_mou_8, loc_og_t2c_mou_6, loc_og_t2c_mou_7, loc_og_t2c_mou_8, loc_og_mou_6, loc_og_mou_7, loc_og_mou_8, std_og_t2t_mou_6, std_og_t2t_mou_7, std_og_t2t_mou_8, std_og_t2m_mou_6, std_og_t2m_mou_7, std_og_t2m_mou_8, std_og_t2f_mou_6, std_og_t2f_mou_7, std_og_t2f_mou_8, std_og_t2c_mou_6, std_og_t2c_mou_7, std_og_t2c_mou_8, std_og_mou_6, std_og_mou_7, std_og_mou_8, isd_og_mou_6, isd_og_mou_7, isd_og_mou_8, spl_og_mou_6, spl_og_mou_7, spl_og_mou_8, og_others_6, og_others_7, og_others_8, total_og_mou_6, total_og_mou_7, total_og_mou_8, loc_ic_t2t_mou_6, loc_ic_t2t_mou_7, loc_ic_t2t_mou_8, loc_ic_t2m_mou_6, loc_ic_t2m_mou_7, loc_ic_t2m_mou_8, loc_ic_t2f_mou_6, loc_ic_t2f_mou_7, loc_ic_t2f_mou_8, loc_ic_mou_6, loc_ic_mou_7, loc_ic_mou_8, std_ic_t2t_mou_6, std_ic_t2t_mou_7, std_ic_t2t_mou_8, std_ic_t2m_mou_6, std_ic_t2m_mou_7, std_ic_t2m_mou_8, std_ic_t2f_mou_6, std_ic_t2f_mou_7, std_ic_t2f_mou_8, std_ic_t2o_mou_6, std_ic_t2o_mou_7, std_ic_t2o_mou_8, std_ic_mou_6, std_ic_mou_7, std_ic_mou_8, total_ic_mou_6, total_ic_mou_7, total_ic_mou_8, spl_ic_mou_6, spl_ic_mou_7, spl_ic_mou_8, isd_ic_mou_6, isd_ic_mou_7, ...]
Index: []

[0 rows x 172 columns]

In [337]:
# List of variables names with NA or NaN values 
df.loc[:,df.isna().any(axis=0)].columns

Index(['loc_og_t2o_mou', 'std_og_t2o_mou', 'loc_ic_t2o_mou',
       'last_date_of_month_7', 'last_date_of_month_8', 'onnet_mou_6',
       'onnet_mou_7', 'onnet_mou_8', 'offnet_mou_6', 'offnet_mou_7',
       ...
       'arpu_3g_8', 'arpu_2g_6', 'arpu_2g_7', 'arpu_2g_8', 'night_pck_user_6',
       'night_pck_user_7', 'night_pck_user_8', 'fb_user_6', 'fb_user_7',
       'fb_user_8'],
      dtype='object', length=125)

##### Observation : There are around 125 features conatining NaN values in it.

In [338]:
df.loc[:,df.isna().any(axis=0)]

loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou last_date_of_month_7  \
0                 0.0             0.0             0.0            7/31/2014   
1                 0.0             0.0             0.0            7/31/2014   
2                 0.0             0.0             0.0            7/31/2014   
3                 0.0             0.0             0.0            7/31/2014   
4                 0.0             0.0             0.0            7/31/2014   
...               ...             ...             ...                  ...   
69994             0.0             0.0             0.0            7/31/2014   
69995             0.0             0.0             0.0            7/31/2014   
69996             0.0             0.0             0.0            7/31/2014   
69997             0.0             0.0             0.0            7/31/2014   
69998             0.0             0.0             0.0            7/31/2014   

      last_date_of_month_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  \
0                8/31/2014        48.58       124.38         1.29   
1                8/31/2014         0.00         0.00         0.00   
2                8/31/2014         0.53        15.93         0.00   
3                8/31/2014         7.26        16.01         0.00   
4                8/31/2014        21.28         4.83         6.13   
...                    ...          ...          ...          ...   
69994            8/31/2014         0.00         7.36        10.93   
69995            8/31/2014       116.54       196.46       232.63   
69996            8/31/2014        77.13        68.44        78.44   
69997            8/31/2014        14.01         7.64         6.71   
69998            8/31/2014         0.00         0.00         0.00   

       offnet_mou_6  offnet_mou_7  ...  arpu_3g_8  arpu_2g_6  arpu_2g_7  \
0             32.24         96.68  ...        NaN        NaN        NaN   
1              0.00         25.99  ...        NaN        NaN     122.08   
2             53.99         82.05  ...        NaN        NaN        NaN   
3             68.76         78.48  ...        NaN        NaN        NaN   
4             56.99         38.11  ...       0.00       0.00      35.12   
...             ...           ...  ...        ...        ...        ...   
69994          0.00        488.46  ...       0.00        NaN      60.61   
69995         49.53         96.28  ...        NaN        NaN        NaN   
69996        335.54        227.94  ...        NaN        NaN        NaN   
69997         30.34         16.68  ...     122.08     150.67     212.17   
69998          0.00          0.00  ...       0.00       0.00        NaN   

       arpu_2g_8  night_pck_user_6  night_pck_user_7  night_pck_user_8  \
0            NaN               NaN               NaN               NaN   
1            NaN               NaN               0.0               NaN   
2            NaN               NaN               NaN               NaN   
3            NaN               NaN               NaN               NaN   
4           0.00               0.0               0.0               0.0   
...          ...               ...               ...               ...   
69994      48.00               NaN               0.0               0.0   
69995        NaN               NaN               NaN               NaN   
69996        NaN               NaN               NaN               NaN   
69997     122.07               0.0               0.0               0.0   
69998       0.02               0.0               NaN               0.0   

       fb_user_6  fb_user_7  fb_user_8  
0            NaN        NaN        NaN  
1            NaN        1.0        NaN  
2            NaN        NaN        NaN  
3            NaN        NaN        NaN  
4            1.0        1.0        1.0  
...          ...        ...        ...  
69994        NaN        1.0        1.0  
69995        NaN        NaN        NaN  
69996        NaN        NaN        NaN  
69997        1.0        1.0        1.0  


In [339]:
# checking percentage of null values in  each feature of train set
df_null=df[df.columns[df.isnull().any()]].isnull().sum() * 100 / df.shape[0]
df_null

loc_og_t2o_mou           1.002871
std_og_t2o_mou           1.002871
loc_ic_t2o_mou           1.002871
last_date_of_month_7     0.570008
last_date_of_month_8     1.047158
                          ...    
night_pck_user_7        74.478207
night_pck_user_8        73.689624
fb_user_6               74.902499
fb_user_7               74.478207
fb_user_8               73.689624
Length: 125, dtype: float64

In [340]:
# checking percentage of null values in  each feature of test set
df_null_test=df[df.columns[df.isnull().any()]].isnull().sum() * 100 / df.shape[0]
df_null_test

loc_og_t2o_mou           1.002871
std_og_t2o_mou           1.002871
loc_ic_t2o_mou           1.002871
last_date_of_month_7     0.570008
last_date_of_month_8     1.047158
                          ...    
night_pck_user_7        74.478207
night_pck_user_8        73.689624
fb_user_6               74.902499
fb_user_7               74.478207
fb_user_8               73.689624
Length: 125, dtype: float64

In [341]:
# features before dropping value having null value more than 30% 
for i in df.columns:
    print(i)

id
circle_id
loc_og_t2o_mou
std_og_t2o_mou
loc_ic_t2o_mou
last_date_of_month_6
last_date_of_month_7
last_date_of_month_8
arpu_6
arpu_7
arpu_8
onnet_mou_6
onnet_mou_7
onnet_mou_8
offnet_mou_6
offnet_mou_7
offnet_mou_8
roam_ic_mou_6
roam_ic_mou_7
roam_ic_mou_8
roam_og_mou_6
roam_og_mou_7
roam_og_mou_8
loc_og_t2t_mou_6
loc_og_t2t_mou_7
loc_og_t2t_mou_8
loc_og_t2m_mou_6
loc_og_t2m_mou_7
loc_og_t2m_mou_8
loc_og_t2f_mou_6
loc_og_t2f_mou_7
loc_og_t2f_mou_8
loc_og_t2c_mou_6
loc_og_t2c_mou_7
loc_og_t2c_mou_8
loc_og_mou_6
loc_og_mou_7
loc_og_mou_8
std_og_t2t_mou_6
std_og_t2t_mou_7
std_og_t2t_mou_8
std_og_t2m_mou_6
std_og_t2m_mou_7
std_og_t2m_mou_8
std_og_t2f_mou_6
std_og_t2f_mou_7
std_og_t2f_mou_8
std_og_t2c_mou_6
std_og_t2c_mou_7
std_og_t2c_mou_8
std_og_mou_6
std_og_mou_7
std_og_mou_8
isd_og_mou_6
isd_og_mou_7
isd_og_mou_8
spl_og_mou_6
spl_og_mou_7
spl_og_mou_8
og_others_6
og_others_7
og_others_8
total_og_mou_6
total_og_mou_7
total_og_mou_8
loc_ic_t2t_mou_6
loc_ic_t2t_mou_7
loc_ic_t2t_mou_8
loc

In [342]:
# Dropping features more than 30% null values in train set
df = df.drop(df.columns[df.apply(lambda col: col.isnull().sum()/len(df) > 0.30)], axis=1)

In [343]:
# features after dropping value having null value more than 30%.
for i in df.columns:
    print(i)

id
circle_id
loc_og_t2o_mou
std_og_t2o_mou
loc_ic_t2o_mou
last_date_of_month_6
last_date_of_month_7
last_date_of_month_8
arpu_6
arpu_7
arpu_8
onnet_mou_6
onnet_mou_7
onnet_mou_8
offnet_mou_6
offnet_mou_7
offnet_mou_8
roam_ic_mou_6
roam_ic_mou_7
roam_ic_mou_8
roam_og_mou_6
roam_og_mou_7
roam_og_mou_8
loc_og_t2t_mou_6
loc_og_t2t_mou_7
loc_og_t2t_mou_8
loc_og_t2m_mou_6
loc_og_t2m_mou_7
loc_og_t2m_mou_8
loc_og_t2f_mou_6
loc_og_t2f_mou_7
loc_og_t2f_mou_8
loc_og_t2c_mou_6
loc_og_t2c_mou_7
loc_og_t2c_mou_8
loc_og_mou_6
loc_og_mou_7
loc_og_mou_8
std_og_t2t_mou_6
std_og_t2t_mou_7
std_og_t2t_mou_8
std_og_t2m_mou_6
std_og_t2m_mou_7
std_og_t2m_mou_8
std_og_t2f_mou_6
std_og_t2f_mou_7
std_og_t2f_mou_8
std_og_t2c_mou_6
std_og_t2c_mou_7
std_og_t2c_mou_8
std_og_mou_6
std_og_mou_7
std_og_mou_8
isd_og_mou_6
isd_og_mou_7
isd_og_mou_8
spl_og_mou_6
spl_og_mou_7
spl_og_mou_8
og_others_6
og_others_7
og_others_8
total_og_mou_6
total_og_mou_7
total_og_mou_8
loc_ic_t2t_mou_6
loc_ic_t2t_mou_7
loc_ic_t2t_mou_8
loc

In [344]:
# 30 features had null values more than 30% so dropped it.

In [345]:
df.shape

(69999, 142)

In [346]:
# Dropping features more than 30% null values in test set
df_test = df_test.drop(df_test.columns[df_test.apply(lambda col: col.isnull().sum()/len(df_test) > 0.30)], axis=1)
df_test.shape

(30000, 141)

#### Observation : Total 30 features having null values more than 30% in test set, dropped it

In [347]:
# dropping features have only one value in column or have only one unique value in train set
for i in df.columns:
    if df[i].nunique() == 1:
        df.drop(i,axis=1,inplace = True)
df.shape

(69999, 129)

#### Observation : Total 13 features with unique value in train set, dropped it

In [348]:
# dropping features have only one value in column or have only one unique value
for i in df_test.columns:
    if df_test[i].nunique() == 1:
        df_test.drop(i,axis=1,inplace = True)
df_test.shape

(30000, 128)

#### Observation  : Total 13 features with unique value in test set, dropped it

In [349]:
# Count the rows having more than 50% missing values
df_missing_rows = df[(df.isnull().sum(axis=1)) > (len(df.columns)//2)]
df_missing_rows.shape

(758, 129)

#### Observation : 758 rows have missing values in train set

In [350]:
df = df.drop(df_missing_rows.index)
df.shape

(69241, 129)

#### Observation : After dropping rows with missing values more than 50%, 69241 rows remains in train set

In [351]:
# checking count of null value of each feature in train set.
df.isnull().sum()

id                      0
arpu_6                  0
arpu_7                  0
arpu_8                  0
onnet_mou_6          2010
                     ... 
aon                     0
aug_vbc_3g              0
jul_vbc_3g              0
jun_vbc_3g              0
churn_probability       0
Length: 129, dtype: int64

In [352]:
# checking count of null value of each feature in test set.
df_test.isnull().sum()

id                0
arpu_6            0
arpu_7            0
arpu_8            0
onnet_mou_6    1169
               ... 
sachet_3g_8       0
aon               0
aug_vbc_3g        0
jul_vbc_3g        0
jun_vbc_3g        0
Length: 128, dtype: int64

In [353]:
# imputing median value in place of null values in train set
df.fillna(df[df.columns].median(),inplace=True)

In [354]:
# imputing median value in place of null values in test set
df_test.fillna(df_test[df_test.columns].median(),inplace=True)

In [355]:
df.isnull().sum()

id                   0
arpu_6               0
arpu_7               0
arpu_8               0
onnet_mou_6          0
                    ..
aon                  0
aug_vbc_3g           0
jul_vbc_3g           0
jun_vbc_3g           0
churn_probability    0
Length: 129, dtype: int64

In [356]:
# no null values after imputting median value in train set

In [357]:
df_test.isnull().sum()

id             0
arpu_6         0
arpu_7         0
arpu_8         0
onnet_mou_6    0
              ..
sachet_3g_8    0
aon            0
aug_vbc_3g     0
jul_vbc_3g     0
jun_vbc_3g     0
Length: 128, dtype: int64

In [358]:
# no null values after imputting median value in test set

In [359]:
df.columns

Index(['id', 'arpu_6', 'arpu_7', 'arpu_8', 'onnet_mou_6', 'onnet_mou_7',
       'onnet_mou_8', 'offnet_mou_6', 'offnet_mou_7', 'offnet_mou_8',
       ...
       'monthly_3g_7', 'monthly_3g_8', 'sachet_3g_6', 'sachet_3g_7',
       'sachet_3g_8', 'aon', 'aug_vbc_3g', 'jul_vbc_3g', 'jun_vbc_3g',
       'churn_probability'],
      dtype='object', length=129)

In [360]:
# Dropping ID column from both Train as well as Test Data

df.drop(columns = 'id', axis = 1, inplace = True)
df_test.drop(columns = 'id', axis = 1, inplace = True)

In [361]:
# Calculating Churn Probability
df['churn_probability'].value_counts()/len(df)*100

0    90.478185
1     9.521815
Name: churn_probability, dtype: float64

In [362]:
# Finding Date column list
date_cols = [k for k in df.columns.to_list() if 'date' in k]
print(date_cols) 

['date_of_last_rech_6', 'date_of_last_rech_7', 'date_of_last_rech_8']


In [363]:
# Date columns doesn't have sufficient amount of data so dropping it from train set
df = df.drop(date_cols, axis=1)

In [364]:
# Date columns doesn't have sufficient amount of data so dropping it from test set
df_test = df_test.drop(date_cols, axis=1)

In [365]:
# Printing shape of train and test data
print(df.shape)
print(df_test.shape)

(69241, 125)
(30000, 124)


In [366]:
# Handling outliers of numerical variables by taking upper_quantile=0.99 and lower_quantile=0.25
def drop_outliers(x):
    lists = []
    for col in df.columns:
        Q1 = x[col].quantile(.25)
        Q3 = x[col].quantile(.99)
        IQR = Q3-Q1
        x =  x[(x[col] >= (Q1-(1.5*IQR))) & (x[col] <= (Q3+(1.5*IQR)))] 
    return x   

df = drop_outliers(df)

In [367]:
# Printing shape of train and test data
print(df.shape)
print(df_test.shape)

(61009, 125)
(30000, 124)


#### Observation : after removing outliers from train set we can observe reduction in rows

## EDA (Exploratory Data Analysis)

In [ ]:
# Printing box plot of all the columns with repect to "churn_probability"
for col in df.columns:
    plt.figure()
    sns.boxplot(x=col, y='churn_probability', data=df)
    plt.show()

In [ ]:
# bar plot of important feature with target variable.
plt.figure()
sns.barplot(x='churn_probability', y='total_og_mou_6', data=df)
plt.figure()
sns.barplot(x='churn_probability', y='total_og_mou_7', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='total_og_mou_8', data=df)
plt.figure()
sns.barplot(x='churn_probability', y='onnet_mou_6', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='onnet_mou_7', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='onnet_mou_8', data=df)
plt.figure()

### Observations

#### Due to Outgoing voice calls , churn is maximum in month_6  then month_7 and month_8 it decreases fastly
#### Probability of being churn is less in outgoing month_7 as well in month_8
#### Probability of being churn is more in onnet voice call month_6 and less in month_8
#### Probability of being churn or non_churn is same for onnet voice call month_7

In [ ]:
plt.figure()
sns.barplot(x='churn_probability', y='total_rech_num_6', data=df)
plt.figure()
sns.barplot(x='churn_probability', y='total_rech_num_7', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='total_rech_num_8', data=df)
plt.figure()
sns.barplot(x='churn_probability', y='total_rech_amt_6', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='total_rech_amt_7', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='total_rech_amt_8', data=df)
plt.figure()

### Observations

#### Probability of being churn is less for the customers who recharge or recharge more amount in month_8

In [ ]:
plt.figure()
sns.barplot(x='churn_probability', y='aon', data=df)
plt.show()

### Observation
#### If operator is using same network(T) from many days then its probabilty to churn is less.

In [ ]:
plt.figure()
sns.boxplot(x='churn_probability', y='aon', data=df)
plt.show()

### Observation
#### Mean amount of churn is equal to 25% of non-churn

In [ ]:
plt.figure()
sns.barplot(x='churn_probability', y='vol_2g_mb_6', data=df)
plt.figure()
sns.barplot(x='churn_probability', y='vol_2g_mb_7', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='vol_2g_mb_8', data=df)
plt.figure()
sns.barplot(x='churn_probability', y='vol_3g_mb_6', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='vol_3g_mb_7', data=df)
plt.show()
plt.figure()
sns.barplot(x='churn_probability', y='vol_3g_mb_8', data=df)
plt.figure()

### Observation
#### Customer with more usage of 2g and 3g data churns less with 8 month customer

In [ ]:
# Printing shape of train and test data
print(df.shape)
print(df_test.shape)

## Data preprocessing

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [ ]:
# Dividing into X and y sets for the model building
X = df.drop('churn_probability', axis = 1)
y = df['churn_probability']

In [ ]:
# Checking shape of X and Y set
print(X.shape)
y.shape

## Train-test split

In [ ]:
# splited data into train and test results in train data containing 75% of data and test data containing 25% of data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 100)

In [ ]:
# Checking shape of X-test set
X_test.shape

In [ ]:
# Checking shape of X-train set
X_train.shape

In [ ]:
# Checking shape of y-test set
y_train.shape

# Scaling features

In [ ]:
# Rescaling the Features
# We will use MinMax scaling.
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

In [ ]:
# scaling train and test set
X_train_sc = scaler.fit_transform(X_train)
X_test_sc = scaler.transform(X_test)

In [ ]:
# Checking shape of X-train set after fit transformation
X_train_sc.shape

In [ ]:
# Checking shape of X-test set after transformation
X_test_sc.shape

In [ ]:
# Checking data of X-train set after fit transformation
X_train_sc

In [ ]:
# Checking data of Y-test set after  transformation
X_test_sc

## PCA (Principal Component Analysis)

In [ ]:
# Importing PCA module
# fitting PCA on scaled training set
from sklearn.decomposition import PCA
pca = PCA(random_state=42)
pca.fit(X_train_sc)

In [ ]:
# Checking pca components value after PCA fit on train-set 
pca.components_

In [ ]:
# Checking pca components variance after PCA fit on train-set 
pca.explained_variance_ratio_

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
# Visualisation of no. of PCs with variance
plt.bar(range(1,len(pca.explained_variance_ratio_)+1), pca.explained_variance_ratio_)

#### Observation : PCA_1 component has 15% information(maximum), PCA_2 component has 10% information and has we go further component containing  information decreases.

In [ ]:
var_cumu = np.cumsum(pca.explained_variance_ratio_)

In [ ]:
# Scree plot to find the no. of PC required to get desired value of variance
plt.plot(range(1,len(var_cumu)+1), var_cumu)

In [ ]:
# Calculating the Principle components for variance cover of 95%
pca_avg = PCA(0.95)

In [ ]:
#Applying PCA components to train set
df_train_pca = pca_avg.fit_transform(X_train_sc)
df_train_pca.shape

In [ ]:
# Applying the PCA components to Test Set

df_test_pca = pca_avg.transform(X_test_sc)
df_test_pca.shape

#### Observation : after performing dimensional reduction variables decrease to 64 on both train and test set

In [ ]:
# checking correlation between features after performing PCA on train set
corr = np.corrcoef(df_train_pca.transpose())
plt.figure(figsize = (30, 30))
sns.heatmap(corr)
plt.show()

In [ ]:
# checking correlation between features after performing PCA on test set
corr = np.corrcoef(df_test_pca.transpose())
plt.figure(figsize = (30, 30))
sns.heatmap(corr)
plt.show()

#### Observation : correlation handled by PCA(no correlation between features) on both train and test set

In [ ]:
#Checking shape of PCs of train set 
df_train_pca.shape

In [ ]:
#Checking shape of PCs of test set 
df_test_pca.shape

## Logistic regression with PCA

In [ ]:
# Importing different module
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV, RepeatedStratifiedKFold, StratifiedKFold, RandomizedSearchCV
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.experimental import enable_halving_search_cv
from sklearn.model_selection import HalvingGridSearchCV
from sklearn.feature_selection import RFE

In [ ]:
# Assigning no. of split for the fold
fold = StratifiedKFold(n_splits = 5)

In [ ]:
# Creating a logistic regression model 
grid_lr = [
    {    "penalty":['l1','l2'],
        "class_weight":[{0:1,1:1}, {0:1,1:2}, {0:1,1:3}]
    }    
]

gscv = GridSearchCV(LogisticRegression(solver='lbfgs'), grid_lr, cv = fold , scoring = 'accuracy', verbose = 1, n_jobs = -1)

In [ ]:
# Fitting the model on trained PCs set
gfit_lr = gscv.fit(df_train_pca, y_train)

In [ ]:
# Checking score and parameter of the model
print("score : {}".format(gfit_lr.best_score_))
logistic_params = gfit_lr.best_params_
print(logistic_params)

#### Observation : accuracy of logistic regression is 91.65% on train set

In [ ]:
# Finding best model
lr_best = gfit_lr.best_estimator_

In [ ]:
# Finding preddiction on the trainset
X_train_pred = lr_best.predict(df_train_pca)

In [ ]:
# Printing coefficient of the best fit model
print(lr_best.coef_.shape)
lr_best.coef_

In [ ]:
# Predicting on PCs test set
y_pred_lr = lr_best.predict(df_test_pca)

In [ ]:
# Printing confusion matrix
print(confusion_matrix(y_test, y_pred_lr))

In [ ]:
#Printing model report on the train and test set after PCA
print(classification_report(y_test, y_pred_lr))

#### Observation : accuracy of logistic regression is 92.00% on test set

## Finding ROC of Logistic + PCA model

In [ ]:
def draw_roc( actual, probs ):
    fpr, tpr, thresholds = metrics.roc_curve( actual, probs,
                                              drop_intermediate = False )
    auc_score = metrics.roc_auc_score( actual, probs )
    plt.figure(figsize=(5, 5))
    plt.plot( fpr, tpr, label='ROC curve (area = %0.2f)' % auc_score )
    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate or [1 - True Negative Rate]')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver operating characteristic example')
    plt.legend(loc="lower right")
    plt.show()

    return None

In [ ]:
from sklearn import metrics

In [ ]:
fpr, tpr, thresholds = metrics.roc_curve( y_train, X_train_pred, drop_intermediate = False )

In [ ]:
draw_roc(y_train, X_train_pred)

#### Observation : ROC is 0.74

## Random forest with PCA

In [ ]:
# Creating a Random Forest model
grid_rd = [
    {   'n_estimators' : [250],
        'max_depth' : [ 7], 
        'max_features' : [ 15],
        'min_samples_split':[3, 4],
        'min_samples_leaf':[1,2],
        "class_weight":['balanced', {0:1,1:10}]
    }    
]

gscv2 = HalvingGridSearchCV(RandomForestClassifier(random_state = 100), grid_rd, cv = fold , scoring = 'accuracy', verbose = 1, n_jobs = -1)

In [ ]:
# Fitting the model on trained PCs set
gfit_rd = gscv2.fit(df_train_pca, y_train)

In [ ]:
# Checking score and parameter of the model
print("score : {}".format(gfit_rd.best_score_))
gfit_rd.best_params_

#### Observation : Accuracy of random forest is 84.88% on train set

In [ ]:
# Finding best model

rf_best = gfit_rd.best_estimator_

In [ ]:
# Finding preddiction on the trainset

X_train_pred_rd = rf_best.predict(df_train_pca)

In [ ]:
# Printing important features

rf_best.feature_importances_

In [ ]:
# Predicting on PCs test set

y_pred_rd = rf_best.predict(df_test_pca)

In [ ]:
# Printing confusion matrix

print(confusion_matrix(y_test, y_pred_rd))

In [ ]:
#Printing model report on the train and test set after PCA

print(classification_report(y_test, y_pred_rd))

### Observation
#### Accuracy of random forest is 85% on test set

## Finding ROC of Randon Forest  + PCA model

In [ ]:
def draw_roc( actual, probs ):
    fpr, tpr, thresholds = metrics.roc_curve( actual, probs,
                                              drop_intermediate = False )
    auc_score = metrics.roc_auc_score( actual, probs )
    plt.figure(figsize=(5, 5))
    plt.plot( fpr, tpr, label='ROC curve (area = %0.2f)' % auc_score )
    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate or [1 - True Negative Rate]')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver operating characteristic example')
    plt.legend(loc="lower right")
    plt.show()

    return None

In [ ]:
fpr, tpr, thresholds = metrics.roc_curve( y_train, X_train_pred_rd, drop_intermediate = False )

In [ ]:
draw_roc(y_train, X_train_pred_rd)

#### Observation : ROC is 0.84

## XG-boost with PCA

In [ ]:
# Creating a XGBoost model 
grid_xgb = [
    { 
        'n_estimators': [1200],
        'learning_rate': [0.01],
        'max_depth': [4,5],
        'min_child_weight': [2],
        "class_weight":[{0:1,1:10}, 'balanced' ]
    }    
]

gscv3 = RandomizedSearchCV(XGBClassifier(random_state = 100, tree_method='gpu_hist', predictor='gpu_predictor'), grid_xgb, cv = fold , scoring = 'accuracy', n_jobs = -1, n_iter=20)

In [ ]:
# Fitting the model on trained PCs set
gfit_xgb = gscv3.fit(df_train_pca, y_train)

In [ ]:
# Checking score and parameter of the model
print("score : {}".format(gfit_xgb.best_score_))
xgbparams = gfit_xgb.best_params_
print(xgbparams)

#### Observation : Accuracy of xg-boost is 92.09% on train set

In [ ]:
xgb_best = gfit_xgb.best_estimator_

In [ ]:
def draw_roc( actual, probs ):
    fpr, tpr, thresholds = metrics.roc_curve( actual, probs,
                                              drop_intermediate = False )
    auc_score = metrics.roc_auc_score( actual, probs )
    plt.figure(figsize=(5, 5))
    plt.plot( fpr, tpr, label='ROC curve (area = %0.2f)' % auc_score )
    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate or [1 - True Negative Rate]')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver operating characteristic example')
    plt.legend(loc="lower right")
    plt.show()

    return None

In [ ]:
# Finding best model
rf_best = gfit_xgb.best_estimator_

In [ ]:
# Finding preddiction on the trainset
X_train_pred_xgb = rf_best.predict(df_train_pca)

In [ ]:
# Printing coefficient of the best fit model feature
xgb_best.feature_importances_

In [ ]:
# Predicting on PCs test set
y_pred_xgb = xgb_best.predict(df_test_pca)

In [ ]:
# Printing confusion matrix
print(confusion_matrix(y_test, y_pred_xgb))

In [ ]:
#Printing model report on the train and test set after PCA
print(classification_report(y_test, y_pred_xgb))

#### Observation : Accuracy of xg-boost is 92.00% on test set

## Finding ROC of XGBoost + PCA model

In [ ]:
fpr, tpr, thresholds = metrics.roc_curve( y_train, X_train_pred_xgb, drop_intermediate = False )

In [ ]:
draw_roc(y_train, X_train_pred_xgb)

#### Observation : ROC is 0.75

## Logistic regression without PCA (using RFE)

In [ ]:
# Importing module
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()

In [ ]:
# performing RFE on 50 features
from sklearn.feature_selection import RFE
rfe = RFE(lr, n_features_to_select=50)            
rfe = rfe.fit(X_train_sc, y_train)

In [ ]:
# Printing all the features with it's rank after RFE
list(zip(X_train.columns, rfe.support_, rfe.ranking_))

In [ ]:
# Listing top 50 columns after RFE
col = X_train.columns[rfe.support_]
col

In [ ]:
# Creating a new X-train with to 50 features after RFE
X_train_rfe = X_train[col]

In [ ]:
# Checking shape of new X-train set
X_train_rfe.shape

In [ ]:
# Hyperparameter tuning with Logistic Regression

In [ ]:
grid_LR = [
    {    "penalty":['l1','l2'],
        "class_weight":[{0:1,1:1}, {0:1,1:2}, {0:1,1:3}]
    }    
]

gscv = GridSearchCV(LogisticRegression(solver='lbfgs'), grid_LR, cv = fold , scoring = 'accuracy', verbose = 1, n_jobs = -1)

In [ ]:
# Fitting the model on train set
gfit_LR = gscv.fit(X_train_rfe, y_train)

In [ ]:
# Checking score and parameter of the model
print("score : {}".format(gfit_LR.best_score_))
logistic_params = gfit_LR.best_params_
print(logistic_params)

#### Observation : accuracy of logistic regression using RfE is 93.30% on train set

In [ ]:
# Finding  best estimator
LR_best = gfit_LR.best_estimator_

In [ ]:
# Finding  best estimator
lr_best_2 = gfit_LR.best_estimator_

In [ ]:
# Predicting on train set
X_train_pred_rfe = lr_best_2.predict(X_train_rfe)

In [ ]:
X_test_new = X_test[X_train_rfe.columns]

In [ ]:
y_pred_lr = lr_best_2.predict(X_test_new)

In [ ]:
print(classification_report(y_test, y_pred_lr))

#### Observation : Accuracy of logistic regression using RFE is 93.00% on test set

## Finding ROC of Logistic + RFE model

In [ ]:
def draw_roc( actual, probs ):
    fpr, tpr, thresholds = metrics.roc_curve( actual, probs,
                                              drop_intermediate = False )
    auc_score = metrics.roc_auc_score( actual, probs )
    plt.figure(figsize=(5, 5))
    plt.plot( fpr, tpr, label='ROC curve (area = %0.2f)' % auc_score )
    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate or [1 - True Negative Rate]')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver operating characteristic example')
    plt.legend(loc="lower right")
    plt.show()

    return None

In [ ]:
fpr, tpr, thresholds = metrics.roc_curve( y_train, X_train_pred_rfe, drop_intermediate = False )

In [ ]:
draw_roc(y_train, X_train_pred_rfe)

#### Observation : ROC is 0.71

In [ ]:
# top 10 features
LR_result = pd.DataFrame(lr_best_2.coef_.reshape(50,1), index = col, columns = ['Coef'])
LR_result.sort_values(by = 'Coef', key = abs, ascending = False).head(10)

## Random forest without PCA (using RFE)

In [ ]:
grid_RD = [
    {   'n_estimators' : [250],
        'max_depth' : [ 7], 
        'max_features' : [20],
        'min_samples_split':[ 4],
    }    
]

gscv2 = RandomizedSearchCV(RandomForestClassifier(random_state = 100, class_weight = 'balanced'), grid_RD, cv = fold , scoring = 'accuracy', verbose = 1, n_jobs = -1)

In [ ]:
# Fitting model on train set
gfit_RD = gscv2.fit(X_train_rfe, y_train)

In [ ]:
# Checking score and parameter of the model
print("score : {}".format(gfit_RD.best_score_))
gfit_RD.best_params_

#### Observation : Accuracy of ranfom forest using RFE is 91.26% on test set

In [ ]:
# Finding best estimator
rd_best_2 = gfit_RD.best_estimator_

In [ ]:
# Predcting on train set
X_train_pred_rfe_rd = rd_best_2.predict(X_train_rfe)

In [ ]:
# Creating new test set based on top RFE columns
X_test_new = X_test[X_train_rfe.columns]

In [ ]:
# Predicting on test set
y_pred_rd = rd_best_2.predict(X_test_new)

In [ ]:
# Printing coefficient of the best fit model
print(classification_report(y_test, y_pred_rd))

#### Observation : Accuracy of random forest using RFE is 91.00% on test set

In [ ]:
def draw_roc( actual, probs ):
    fpr, tpr, thresholds = metrics.roc_curve( actual, probs,
                                              drop_intermediate = False )
    auc_score = metrics.roc_auc_score( actual, probs )
    plt.figure(figsize=(5, 5))
    plt.plot( fpr, tpr, label='ROC curve (area = %0.2f)' % auc_score )
    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate or [1 - True Negative Rate]')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver operating characteristic example')
    plt.legend(loc="lower right")
    plt.show()

    return None

In [ ]:
fpr, tpr, thresholds = metrics.roc_curve( y_train, X_train_pred_rfe_rd, drop_intermediate = False )

In [ ]:
draw_roc(y_train, X_train_pred_rfe_rd)

#### Observation : ROC is 0.89

In [ ]:
# top 20 features

In [ ]:
rf_df = pd.DataFrame({'feature' : X_train_rfe.columns, 'importance' : rd_best_2.feature_importances_})
rf_df = rf_df.sort_values(by = 'importance', ascending = False).head(20)
fig = plt.figure(figsize = (12, 9))
sns.barplot(data = rf_df, y = 'feature', x = 'importance')
plt.show()

## Prediction on test set

In [ ]:
df_test = scaler.transform(df_test)

In [ ]:
df_test_pca = pca_avg.transform(df_test)

In [ ]:
y_pred_final = xgb_best.predict(df_test_pca)

In [ ]:
sample = pd.read_csv('sample.csv')
sample.head()

In [ ]:
sample['churn_probability'] = y_pred_final

In [ ]:
sample['churn_probability'].value_counts()

In [ ]:
submission = sample.to_csv('submissions.csv', index=False)

In [ ]:
sample.head()